File: sdk/go/transport/webrtcws/signaling.go
package webrtcws

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"

	"github.com/babelforce/rtvbp/sdk/go"
	"github.com/google/uuid"
	"github.com/pion/webrtc/v4"
)

const (
	offerMethod       = "transport.webrtc.offer"
	maxSignalFrameLen = 1 << 20
	maxSDPLen         = 512 << 10
)

type sessionDescription struct {
	SDP string `json:"sdp"`
}

func negotiateOffer(ctx context.Context, control rtvbp.ControlChannel, envelope rtvbp.Envelope, peer *webrtc.PeerConnection) error {
	offer, err := peer.CreateOffer(nil)
	if err != nil {
		return fmt.Errorf("create WebRTC offer: %w", err)
	}
	complete := webrtc.GatheringCompletePromise(peer)
	if err := peer.SetLocalDescription(offer); err != nil {
		return fmt.Errorf("set local WebRTC offer: %w", err)
	}
	if err := awaitGathering(ctx, complete); err != nil {
		return err
	}
	local := peer.LocalDescription()
	if local == nil {
		return errors.New("webrtcws: local offer is missing")
	}
	payload, err := encodeDescription(local.SDP)
	if err != nil {
		return err
	}
	id := uuid.NewString()
	encoded, err := envelope.Encode(rtvbp.ControlFrame{
		Kind:    rtvbp.KindRequest,
		ID:      id,
		Method:  offerMethod,
		Payload: payload,
	})
	if err != nil {
		return fmt.Errorf("encode WebRTC offer: %w", err)
	}
	if len(encoded) > maxSignalFrameLen {
		return errors.New("webrtcws: encoded offer exceeds signaling limit")
	}
	if err := control.Send(ctx, encoded); err != nil {
		return fmt.Errorf("send WebRTC offer: %w", err)
	}

	received, err := control.Recv(ctx)
	if err != nil {
		return fmt.Errorf("receive WebRTC answer: %w", err)
	}
	frame, err := decodeSignal(envelope, received.Data)
	if err != nil {
		return fmt.Errorf("decode WebRTC answer: %w", err)
	}
	if frame.Kind != rtvbp.KindResponse || frame.CorrelID != id {
		return fmt.Errorf("webrtcws: unexpected answer frame kind=%d correlation=%q", frame.Kind, frame.CorrelID)
	}
	if frame.Err != nil {
		return fmt.Errorf("webrtcws: offer rejected: %d %s", frame.Err.Code, frame.Err.Message)
	}
	answer, err := decodeDescription(frame.Payload)
	if err != nil {
		return fmt.Errorf("decode WebRTC answer payload: %w", err)
	}
	if err := peer.SetRemoteDescription(webrtc.SessionDescription{Type: webrtc.SDPTypeAnswer, SDP: answer.SDP}); err != nil {
		return fmt.Errorf("set remote WebRTC answer: %w", err)
	}
	return nil
}

func negotiateAnswer(ctx context.Context, control rtvbp.ControlChannel, envelope rtvbp.Envelope, peer *webrtc.PeerConnection) error {
	received, err := control.Recv(ctx)
	if err != nil {
		return fmt.Errorf("receive WebRTC offer: %w", err)
	}
	frame, err := decodeSignal(envelope, received.Data)
	if err != nil {
		return fmt.Errorf("decode WebRTC offer: %w", err)
	}
	if frame.Kind != rtvbp.KindRequest || frame.Method != offerMethod || frame.ID == "" {
		return fmt.Errorf("webrtcws: unexpected offer frame kind=%d method=%q id=%q", frame.Kind, frame.Method, frame.ID)
	}
	offer, err := decodeDescription(frame.Payload)
	if err != nil {
		return fmt.Errorf("decode WebRTC offer payload: %w", err)
	}
	if err := peer.SetRemoteDescription(webrtc.SessionDescription{Type: webrtc.SDPTypeOffer, SDP: offer.SDP}); err != nil {
		return fmt.Errorf("set remote WebRTC offer: %w", err)
	}
	answer, err := peer.CreateAnswer(nil)
	if err != nil {
		return fmt.Errorf("create WebRTC answer: %w", err)
	}
	complete := webrtc.GatheringCompletePromise(peer)
	if err := peer.SetLocalDescription(answer); err != nil {
		return fmt.Errorf("set local WebRTC answer: %w", err)
	}
	if err := awaitGathering(ctx, complete); err != nil {
		return err
	}
	local := peer.LocalDescription()
	if local == nil {
		return errors.New("webrtcws: local answer is missing")
	}
	payload, err := encodeDescription(local.SDP)
	if err != nil {
		return err
	}
	encoded, err := envelope.Encode(rtvbp.ControlFrame{
		Kind:     rtvbp.KindResponse,
		CorrelID: frame.ID,
		Payload:  payload,
	})
	if err != nil {
		return fmt.Errorf("encode WebRTC answer: %w", err)
	}
	if len(encoded) > maxSignalFrameLen {
		return errors.New("webrtcws: encoded answer exceeds signaling limit")
	}
	if err := control.Send(ctx, encoded); err != nil {
		return fmt.Errorf("send WebRTC answer: %w", err)
	}
	return nil
}

func awaitGathering(ctx context.Context, complete <-chan struct{}) error {
	select {
	case <-complete:
		return nil
	case <-ctx.Done():
		return fmt.Errorf("gather WebRTC ICE candidates: %w", ctx.Err())
	}
}

func decodeSignal(envelope rtvbp.Envelope, encoded []byte) (rtvbp.ControlFrame, error) {
	if len(encoded) == 0 || len(encoded) > maxSignalFrameLen {
		return rtvbp.ControlFrame{}, errors.New("webrtcws: signaling frame size is invalid")
	}
	return envelope.Decode(encoded)
}

func encodeDescription(sdp string) (json.RawMessage, error) {
	if len(sdp) == 0 || len(sdp) > maxSDPLen {
		return nil, errors.New("webrtcws: SDP size is invalid")
	}
	payload, err := json.Marshal(sessionDescription{SDP: sdp})
	if err != nil {
		return nil, fmt.Errorf("encode WebRTC SDP: %w", err)
	}
	return payload, nil
}

func decodeDescription(payload json.RawMessage) (sessionDescription, error) {
	var description sessionDescription
	if err := json.Unmarshal(payload, &description); err != nil {
		return sessionDescription{}, err
	}
	if len(description.SDP) == 0 || len(description.SDP) > maxSDPLen {
		return sessionDescription{}, errors.New("webrtcws: SDP size is invalid")
	}
	return description, nil
}

File: sdk/rust/src/transport/webrtcws/signaling.rs
use std::sync::Arc;
use std::sync::atomic::{AtomicU64, Ordering};

use serde::{Deserialize, Serialize};
use webrtc::peer_connection::RTCPeerConnection;
use webrtc::peer_connection::sdp::session_description::RTCSessionDescription;

use crate::{ControlChannel, ControlFrame, Envelope, FrameKind};

pub(super) const OFFER_METHOD: &str = "transport.webrtc.offer";
const MAX_SIGNAL_FRAME_LEN: usize = 1 << 20;
const MAX_SDP_LEN: usize = 512 << 10;
static SIGNAL_IDS: AtomicU64 = AtomicU64::new(1);

#[derive(Serialize, Deserialize)]
struct Description {
    sdp: String,
}

pub(super) async fn negotiate_offer(
    control: Arc<dyn ControlChannel>,
    envelope: Arc<dyn Envelope>,
    peer: &RTCPeerConnection,
) -> Result<(), crate::Error> {
    let offer = peer.create_offer(None).await.map_err(transport_error)?;
    let mut gathering = peer.gathering_complete_promise().await;
    peer.set_local_description(offer)
        .await
        .map_err(transport_error)?;
    // The promise completes by either sending or dropping its one-shot sender.
    let _ = gathering.recv().await;
    let local = peer
        .local_description()
        .await
        .ok_or_else(|| crate::Error::Transport("local WebRTC offer is missing".to_owned()))?;
    let payload = encode_description(&local.sdp)?;
    let id = format!(
        "webrtc-offer-{}",
        SIGNAL_IDS.fetch_add(1, Ordering::Relaxed)
    );
    let encoded = envelope.encode(&ControlFrame::request(
        id.clone(),
        OFFER_METHOD,
        Some(payload),
    ))?;
    validate_signal(&encoded)?;
    control.send(encoded).await?;

    let received = control.recv().await?;
    let frame = decode_signal(envelope.as_ref(), &received.data)?;
    if frame.kind != FrameKind::Response || frame.correlation_id != id {
        return Err(crate::Error::Transport(format!(
            "unexpected WebRTC answer kind={:?} correlation={:?}",
            frame.kind, frame.correlation_id
        )));
    }
    if let Some(error) = frame.error {
        return Err(crate::Error::Remote(error));
    }
    let answer = decode_description(frame.payload)?;
    peer.set_remote_description(RTCSessionDescription::answer(answer.sdp).map_err(transport_error)?)
        .await
        .map_err(transport_error)
}

pub(super) async fn negotiate_answer(
    control: Arc<dyn ControlChannel>,
    envelope: Arc<dyn Envelope>,
    peer: &RTCPeerConnection,
) -> Result<(), crate::Error> {
    let received = control.recv().await?;
    let frame = decode_signal(envelope.as_ref(), &received.data)?;
    if frame.kind != FrameKind::Request || frame.method != OFFER_METHOD || frame.id.is_empty() {
        return Err(crate::Error::Transport(format!(
            "unexpected WebRTC offer kind={:?} method={:?} id={:?}",
            frame.kind, frame.method, frame.id
        )));
    }
    let offer = decode_description(frame.payload)?;
    peer.set_remote_description(RTCSessionDescription::offer(offer.sdp).map_err(transport_error)?)
        .await
        .map_err(transport_error)?;
    let answer = peer.create_answer(None).await.map_err(transport_error)?;
    let mut gathering = peer.gathering_complete_promise().await;
    peer.set_local_description(answer)
        .await
        .map_err(transport_error)?;
    let _ = gathering.recv().await;
    let local = peer
        .local_description()
        .await
        .ok_or_else(|| crate::Error::Transport("local WebRTC answer is missing".to_owned()))?;
    let payload = encode_description(&local.sdp)?;
    let encoded = envelope.encode(&ControlFrame::response(frame.id, Some(payload), None))?;
    validate_signal(&encoded)?;
    control.send(encoded).await
}

fn encode_description(sdp: &str) -> Result<serde_json::Value, crate::Error> {
    validate_sdp(sdp)?;
    serde_json::to_value(Description {
        sdp: sdp.to_owned(),
    })
    .map_err(crate::Error::envelope)
}

fn decode_description(payload: Option<serde_json::Value>) -> Result<Description, crate::Error> {
    let description: Description = serde_json::from_value(
        payload
            .ok_or_else(|| crate::Error::Transport("WebRTC SDP payload is missing".to_owned()))?,
    )
    .map_err(crate::Error::envelope)?;
    validate_sdp(&description.sdp)?;
    Ok(description)
}

fn decode_signal(envelope: &dyn Envelope, encoded: &[u8]) -> Result<ControlFrame, crate::Error> {
    validate_signal(encoded)?;
    envelope.decode(encoded)
}

fn validate_signal(encoded: &[u8]) -> Result<(), crate::Error> {
    if encoded.is_empty() || encoded.len() > MAX_SIGNAL_FRAME_LEN {
        Err(crate::Error::Transport(
            "WebRTC signaling frame size is invalid".to_owned(),
        ))
    } else {
        Ok(())
    }
}

fn validate_sdp(sdp: &str) -> Result<(), crate::Error> {
    if sdp.is_empty() || sdp.len() > MAX_SDP_LEN {
        Err(crate::Error::Transport(
            "WebRTC SDP size is invalid".to_owned(),
        ))
    } else {
        Ok(())
    }
}

fn transport_error(error: impl std::fmt::Display) -> crate::Error {
    crate::Error::Transport(error.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn signal_and_sdp_bounds_are_fail_closed() {
        assert!(validate_signal(&[]).is_err());
        assert!(validate_signal(&vec![0; MAX_SIGNAL_FRAME_LEN + 1]).is_err());
        assert!(validate_sdp("").is_err());
        assert!(validate_sdp(&"x".repeat(MAX_SDP_LEN + 1)).is_err());
        assert!(decode_description(Some(serde_json::json!({"sdp": ""}))).is_err());
        assert!(decode_description(Some(serde_json::json!({"other": "x"}))).is_err());
    }
}
